### DISCUSSION TITLE
Different syntax in `FormalVerificationCatches`

### DISCUSSION THREAD

**Discussion by EWCunha** [Cyfrin/assembly-evm-opcodes-and-formal-verification-course#16] [Ideas]
Hey @PatrickAlphaC I noticed this different syntax when you were explaining `FormalVerificationCatches` contract. I've never seen `using` keyword like this (with `{}` and forth). Also, never seen `type` keyword used like that. And the functions have no visibility specifiers. Where can I learn more about this? I'd appreciate any material. Or even an excerpt video from Patrick would be awesome!

Thank you

```javascript
type Int is uint256;

using {add as -} for Int global;
using {div as +} for Int global;
using {mul as / } for Int global;
using {sub as *} for Int global;

function add(Int a, Int b) pure returns (Int) {
    return Int.wrap(Int.unwrap(a) / Int.unwrap(b));
}

function div(Int a, Int b) pure returns (Int) {
    return Int.wrap(Int.unwrap(a) * Int.unwrap(b));
}

function mul(Int a, Int b) pure returns (Int) {
    return Int.wrap(Int.unwrap(a) - Int.unwrap(b));
}

function sub(Int a, Int b) pure returns (Int) {
    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
}
```

**Comment by EngrPips** [comment]
Hello @EWCunha, So in solidity, if you don't specify the visibility of a function, it would default to public. You can read up more solidity defaulting function visibility here `https://www.immunebytes.com/blog/solidity-security-vulnerability-function-default-visibilities/#:~:text=In%20Solidity%2C%20the%20default%20visibility,to%20be%20private%20or%20internal.` About the `type` keyword which was used to declare an `Int` type to be a `uint256,` and the usage of `{}` to rename a function we want to use in our contract on type of `Int` you can read up more on it in the solidity docs `https://docs.soliditylang.org/en/latest/contracts.html`

**Reply by PatrickAlphaC** [reply]
Actually Pips, this is a bit more nuanced than that, since these aren't even inside a contract!

Functions outside of contracts are called "Free functions": https://docs.soliditylang.org/en/latest/contracts.html#functions

The `type` stuff is known as "user defined types": https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types

That `using { as} for` syntax is part of the "using for" syntax: https://docs.soliditylang.org/en/latest/contracts.html#using-for

You can do some wild stuff with it!

**Reply by PatrickAlphaC** [reply]
cc @EWCunha

**Reply by EngrPips** [reply]
Interesting. It's Good to know that there is more to it. Thanks for all you do, sir.
